### ISSUE TITLE
deepcopy of jax.numpy array generates regular numpy array

### ISSUE THREAD

**Issue by durian888** [google/jax#2632]
While updating some existing code that used copy.deepcopy I got what was a rather unexpected result namely that deepcopy returns a regular numpy array when I passed  in a jax.numpy array.  Is this expected behaviour? If so perhaps a note in the docs to indicated this would be good. 

The following is a snippet that replicates the problem.

for `jax.version.__version__ == '0.1.62''`

import jax.numpy as np

import copy

array_in = np.array([1,2,3], dtype= np.int32)
array_out = copy.deepcopy(array_in)

print("type in:" + str(type(array_in))+ "type out:" +str(type(array_out)))

give 
type in:<class 'jax.interpreters.xla.DeviceArray'>type out:<class 'numpy.ndarray'>

**Comment by mattjj** [comment]
Thanks for pointing this out!

AIUI `copy.deepcopy` works by pickling then unpickling an object. We [make JAX DeviceArrays pickle just like NumPy ndarrays](https://github.com/google/jax/blob/52e779c3c18250079d4ee022c36311fdc9531e9c/jax/interpreters/xla.py#L877-L878), and so this deepcopy behavior seems to be a consequence.

While this pickling behavior was the original intent, it might be worth revisiting! @shoyer any thoughts?

(For posterity: [here's](https://stackoverflow.com/a/26599346) how we could restore a DeviceArray.)

**Comment by shoyer** [comment]
If I understand this correctly, that implementation of reduce means that DeviceArray objects unpickle to NumPy arrays, too?

I don't think we want either of these behaviors. Copy, deepcopy and pickle/unpickle should all restore an object with the same type as what you started with.

**Comment by jaivardhankapoor** [comment]
[Sorry for the underspecified question, I am currently writing this on my mobile. Please let me know if you need more info on this query.]

@mattjj  are there any plans to follow-up  on this issue? I am tying to implement a sequential Monte Carlo algorithm with particle resampling, in which I use `deepcopy` in resampling the particle objects. Should I reconvert to `DeviceArray` after resampling the particles or wait for the proper restoration behaviours to be implemented in `JAX`?

**Comment by tawe141** [comment]
I'm developing something that runs into this bug and requires gradient information to be preserved after `deepcopy`s. At the moment, I could probably re-cast everything back to `DeviceArray`s, but it would be great if I didn't have to!

**Comment by shoyer** [comment]
There are at least two surprising behaviors here that should be fixed:

- [ ] Pickling/unpickling a JAX `DeviceArray` objects should return another `DeviceArray`. To do so, we'll need to define our own methods for pickling instead of using NumPy's `__reduce__` directly.
- [ ] Deep-copying JAX arrays should _possibly_ copy device buffers, but certainly should not do so via serializing to the host with pickle. Here we also need to define `__deepcopy__` instead of relying on Python's default behavior. (There is some ambiguousness about whether we need to copy device buffers with deepcopy, given that JAX arrays themselves are immutable, but I think doing so would be within the spirit of deep copy.)

@tawe141 I'm not sure what's going on with example. I believe `deepcopy()` on JAX tracers works just fine (preserving gradients) -- the surprising implementation of copy only arises for arrays that are actually loaded into memory (DeviceArray).

**Comment by JeppeKlitgaard** [comment]
I just wanted to add that I ran into this as well. It is definitely an edge-case, but it took me quite a while to figure out how my arrays suddenly turned into `numpy` arrays (which triggered an error since these don't have the `.at` property).

**Comment by jakevdp** [comment]
A question about design here: where should the deep-copied / unpickled device arrays live? Should they be on the same device as the original? The default device?

This is maybe harder to answer for pickle, because the pickle serialization and de-serialization may happen in entirely different environments.

The easiest behavior would be to strip any device information on serialization, and then de-serialize to the default device

**Comment by JeppeKlitgaard** [comment]
My expectation would be that deepcopy stores the copied array on the same device. Alternatively/additionally one could add a `DeviceArray.deepcopy` method that features a `to` argument:

```python
def deepcopy(self, to: Literal[None, ..., str] = None) -> DeviceArray:
	# None => Same device
	# ... => Default
	# str => Named device
```

For pickling, I think it would make sense to place the array on the default device always.

**Comment by jakevdp** [comment]
I don't think arrays need a `deepcopy()` method; they already have a `copy()` method. We've so far resisted having device specifiers on the jax.numpy API (for example, in functions like `jnp.array` and `jnp.asarray`) but we could revisit that decision.

I don't think the optional argument idea is a good solution here though, because neither `pickle` nor `copy.deepcopy` offer any option to specify extra arguments like that.

**Comment by JeppeKlitgaard** [comment]
I guess since JAX arrays never contain mutable objects (as far as I am aware?), you're right that `deepcopy` would always be the same as `copy`.

One of the use-cases where a "to" copy would be nice is when doing compute on an array on a GPU which is sufficiently large that storing a whole extra copy in VRAM would not be possible (therefore `device_get(arr.copy())` wouldn't work, as far as I can tell, since GPU would go OOM).

I currently have such a use-case where I am periodically saving a potentially large array to later be able to go through the evolution of the array. The current implementation where `copy` leads to an array in host memory works really well in that case.  I am optimistic that this behaviour could still be achieved with `numpy.array(dev_arr, copy=True)` (or possibly even without `copy=True`), which supports not adding a target device argument.

**Comment by jakevdp** [comment]
Remote device transfer is a much deeper issue than a question of keywords to `jax.numpy` operations. We'd first have to get a stable implementation at the low level. There is some functionality for this (see e.g. https://github.com/google/jax/blob/main/tests/remote_transfer_test.py) but it's not stable on all backends.

In any case, I think we'd probably keep this at a lower level even if/when it is available.
